src/Build/Definition/ProjectCollection.cs

```diff
@@ -89,18 +89,29 @@ public class ProjectCollection : IToolsetProvider, IBuildComponent, IDisposable
         private class DisposableReaderWriterLockSlim
         {
             private readonly ReaderWriterLockSlim _lock = new ReaderWriterLockSlim(LockRecursionPolicy.SupportsRecursion);
+
+            // Cache these so that we don't allocate per-call
+            private readonly IDisposable _exitUpgradeableReadLock;
+            private readonly IDisposable _exitWriteLock;
+
+            public DisposableReaderWriterLockSlim()
+            {
+                _exitUpgradeableReadLock = new DelegateDisposable(() => _lock.ExitUpgradeableReadLock());
+                _exitWriteLock = new DelegateDisposable(() => _lock.ExitWriteLock());
+            }
+
             public bool IsWriteLockHeld => _lock.IsWriteLockHeld;
 
             public IDisposable EnterUpgradeableReadLock()
             {
                 _lock.EnterUpgradeableReadLock();
-                return new DelegateDisposable(() => _lock.ExitUpgradeableReadLock());
+                return _exitUpgradeableReadLock;
```

Comment:
Wouldn't a struct disposable achieve the same without allocations?

Reply:
Unlike `foreach`, `using` doesn't support disposal via structs (in my testing on .NET Framework at least). You can make this type a struct, but then it gets boxed (which is an allocation) for the virtual `IDisposable.Dispose` call.

Reply:
`using doesn't support disposal via structs`

I have personally used the pattern, and so does roslyn. See here: https://sourceroslyn.io/#Microsoft.CodeAnalysis/InternalUtilities/SemaphoreSlimExtensions.cs,174ce3d1a7cbf2bb

The idea is to return the struct type, and then ir won't box. I'm not sure if it leaks as an IDisposable in the API, but I just saw this is an internal class.

Reply:
Looks like I was wrong. `using` does support struct disposables without boxing, even on .NET Framework. I'd written an example on sharplab and saw boxing, but I don't have that to hand anymore so don't know what I was doing wrong.

Here's an updated [sharplab example](https://sharplab.io/#v2:EYLgHgbALANALiAlgGwD4AEBMBGAsAKHQAYACdbAOgBUALAJwFMBDAE0QDsBzAbgOLOwBWXvj7YIZTCQDCBAN4ESSgRPRQSAWQAUASkXKF+ZcZIA3JnRIBnBgFsmABxoB7RiQC8JdgwDuJAMp2ji6M/siItlrYOiL6JmSkWjb2Tq4MFAAiiFYOzlZMwMgMAOpMiHC6ekbxJIY1JuQAnFoARMXlNBwkyM4AxgDWLTFxJgC+I+P4kwQccAx07EzIKpIBQamh4bYAomBz7FaIzgfyI+gAzCuBKSEMWTl58yT3ufmFJWUVcJ1WazdpYQi1nWtxgMiY7F6DGQyCYcCO7Cozn6DHYJF6EKhMLhCKRKLRnhYDAAZkwAK7IOBVYx1GrJYJpCilcpaDGQ6Gw+HHPGo4bVeLoADsXl8fwZjBejzoSRBaT5xmm/JIs3mi2WVjgdDJvTgYo2d2yryeIBIAElJW8iiNafEHHREOY5nrboDbCQAPr0/WxJXGC7OtIW+Zaa7ihiu4H/RjUmo2+oer23DyRsMieqK+r+tTPQ2PXQjGkFmrmSyJtJIoPJ03sOZ0HoDBgsCi7Xo0CGcBhaRjEhOyxhg9gU5B8ovxRA9mVRhgV3M2ZW/QcwmPx2qj+PfOjOPzePwAeWAACsGDqgywWwwHFz2FoACQtfzOWwMFx+Ji1RZP5zEkN9g0PGx0DoozziQwAMBwnAkGw/6NiQcA+IgUIUEMaYriQGZoWWjAzjBFAAErQswNi6Kh8STAqBCjEAA===) showing no boxing:

```
        .try
        {
            IL_0015: ldstr "Within lock"
            IL_001a: call void [mscorlib]System.Console::WriteLine(string)
            IL_001f: leave.s IL_002f
        } // end .try
        finally
        {
            // sequence point: hidden
            IL_0021: ldloca.s 0
            IL_0023: constrained. SemaphoreSlimExtensions/SemaphoreDisposer
            IL_0029: callvirt instance void [mscorlib]System.IDisposable::Dispose()
            IL_002e: endfinally
        } // end handler
```

Reply:
I initially thought that [Pattern-based Dispose](https://github.com/dotnet/roslyn/blob/ec6da663c592238cca8e145044e7410c4ca9213a/docs/features/enhanced-using.md) was the problem, but going back even to 2020 build demos, I still can't repro boxing in the IL

Boxing also can be avoided by using `<T>(T target) where T:struct, IDisposable`

Reply:
It might be [this box pattern](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYAYAEMEDoAqAFgE4CmAhgCZgB2A5gNyobYICsTKzCAbNnJgDCqAN6pME1nxgAWTAFkAPPgB8ACnyZg5YnVLAAlJgDuhUmUz4QASQAiYAM4AHAPYPyAIwilxksSklA7Cw1bV19A18gzH9o6JwATjUAIgB1MGBCWkwIFwBjAGtkg044zABfKIrUSq40AUEIcgcHe2c3T29MEEw7R1d3Lx8UWMkYAGZsOTaB0jUjAF4VViSacgBbUhcAMzVG5tb+jqGDEprUVAdgYgBXPOBMAGVru+AZ466evvbB71EqiZTTDvBxzRbLRJqNabHZqZ63e4gzqkU6ccpAA==) that you were thinking of?

It seems like the JIT generates the [same code](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0AKgBYBOApmACYCWAdgOYDc6WuSArCxjnvgEoBXOsBpRy+AMIB7KAAcaAG3KkAyioBuNAMbkI3NquJhScgDJgARgOGjx3VkgBsuBNknoA3umy/sAbQApGmAAcXI6FR0ACmAATzlyaQAzaNVgUkFtYAARGgg5aQgrZQBKUoBdHz8gkPDI0hj4xJToyUVICDyCopLycqq0P3YXOAAWbABZeqjtAB5CAD5owmxgEwZyYFLsAHdiFXJsQhAASW7C4stlat9vIeG/NliNrdLbx/vH76mTAGsVNFStwftgAL4fCEPXwfWphCKzWIJJKpdKZbIXXrXfqVD54UYTaYIxraNFZYALZardakTbbPYHCjHEAQDLkgA02HO+UufQ+Xx+zxpdPe0M+H1Bk3+gOBEr8UOGCr8sMmW2I0iop3kimiquA6s12oA8nJRNI6BB8AA5aSnOiKej0Bi4sX43CE6WkIHYTxQqGsADMrncHQgXR5WOU2BAXMxVxuaAFvjgQfG2Dj5G9AF5FuwAJzROhgcStdqdON9crcf1oVno4DYMkYiPxo4x7k9VtePGpiYZ7O5vAFosl1Fs5udyuytBgoA===) for both anyway.